### PR TITLE
Async events logging, more efficient payload processing

### PR DESCRIPTION
**Pull request by andrewconner**

### CODE REVIEW

File: shoebox/app/com/keepit/common/analytics/PersistEvent.scala [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+case object Load
+case class Update(userId: Id[User])
+case class Persist(event: Event)
+case class PersistMany(events: Seq[Event])
+
+private[analytics] class PersistEventActor extends Actor with Logging {
+
+  def receive() = {
+    case Persist(event) =>
+      log.info("Persisting event %s".format(event.externalId))
```
Comment (eishay) [inline shoebox/app/com/keepit/common/analytics/PersistEvent.scala]: our log is going to be full of those lines very fast, right? do you really need it?

**Comment by andrewconner** [comment]
@eishay Please review. Implemented your suggestions.

File: shoebox/app/com/keepit/common/analytics/PersistEvent.scala [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+      if(diff > 60) {
+        val ex = new Exception("Event log is backing up. Event is %s seconds old".format(diff))
+        inject[HealthcheckPlugin].addError(HealthcheckError(Some(ex), None, None, Healthcheck.INTERNAL, Some(ex.getMessage)))
+        // To keep the event log from backing too far up, ignore very old events.
+        // If we get this, use parallel actors.
+      }
+      else {
+        try { event.persistToS3 } catch { case ex: Throwable =>
+          val ex = new Exception("Could not persist event to S3")
+          inject[HealthcheckPlugin].addError(HealthcheckError(Some(ex), None, None, Healthcheck.INTERNAL, Some(ex.getMessage)))
```
Comment (eishay) [inline shoebox/app/com/keepit/common/analytics/PersistEvent.scala:41]: nice!
You might want to add the event into the error message since the exception may happen when you serialize it.

File: shoebox/app/com/keepit/common/analytics/PersistEvent.scala [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+case class Update(userId: Id[User])
+case class Persist(event: Event)
+case class PersistMany(events: Seq[Event])
+
+private[analytics] class PersistEventActor extends Actor with Logging {
+
+  def receive() = {
+    case Persist(event) =>
+      val diff = Seconds.secondsBetween(event.createdAt, currentDateTime).getSeconds
+      if(diff > 60) {
```
Comment (eishay) [inline shoebox/app/com/keepit/common/analytics/PersistEvent.scala:32]: cool!

File: shoebox/app/com/keepit/controllers/EventController.scala [hunk 4]
```diff
@@ -52,37 +52,41 @@ object EventController extends FortyTwoController {
   private def createEventsFromPayload(params: JsValue, userId: Id[User]) = {
     val logRecievedTime = currentDateTime
 
+    val (user, experiments) = CX.withConnection { implicit conn =>
+      (User.get(userId), UserExperiment.getByUser(userId) map (_.experimentType))
```
Comment (eishay) [inline shoebox/app/com/keepit/controllers/EventController.scala:56]: the AuthenticatedRequest already has `experimants: Seq[State[UserExperiment.ExperimentType]]` you don't need to re-load it from the db

File: shoebox/app/com/keepit/common/analytics/EventLog.scala [hunk 5]
```diff
@@ -62,30 +62,29 @@ trait EventMetadata {
   val prevEvents: Seq[ExternalId[Event]]
 }
 
-case class UserEventMetadata(eventFamily: EventFamily, eventName: String, userId: ExternalId[User], installId: ExternalId[KifiInstallation], userExperiments: Seq[State[UserExperiment.ExperimentType]], metaData: JsValue, prevEvents: Seq[ExternalId[Event]]) extends EventMetadata
+case class UserEventMetadata(eventFamily: EventFamily, eventName: String, userId: ExternalId[User], installId: String, userExperiments: Seq[State[UserExperiment.ExperimentType]], metaData: JsValue, prevEvents: Seq[ExternalId[Event]]) extends EventMetadata
```
Comment (eishay) [inline shoebox/app/com/keepit/common/analytics/EventLog.scala:65]: why did you change it to a string vs a ExternalId[KifiInstallation] ?
